Add solaris build support

epilatow opened the pull request:
A detailed explanation of these changes can be found in issue #1225 

These changes are dependent upon changes to containerd/continuity and
containerd/console, but those changes are not included here.  the lack
of those changes will cause travis ci runs to fail.  those components
must be updated first for travis ci runs with these changes to succeed.

Signed-off-by: Edward Pilatowicz <edward.pilatowicz@oracle.com>

stevvooe commented on client_solaris.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+// DefaultAddress is the default unix socket address
+const DefaultAddress = "/var/run/containerd/containerd.sock"
```
Want to just move the definition? The dial functions look the same.

epilatow replied:
the only difference between client_unix.go and client_solaris.go is DefaultAddress.  i could create client_address_unix.go and client_address_solaris.go, move DefaultAddress to there, and then compile client_unix.go on solaris.  lemme know if that's what you're thinking.

stevvooe replied:
Yes, it would be better to avoid completely copying this code.

stevvooe commented on mount/mount_solaris.go:
```diff
@@ -0,0 +1,55 @@
+	mountCmd = "/usr/sbin/mount"
+)
+
+func doMount(arg ...string) error {
```
What changed here? This looks the same.

epilatow replied:
i'm confused.  i created this file.  what does it look the same as?

stevvooe replied:
This looks the same as the linux version.

epilatow replied:
i've enhanced the comments here to discuss the issues related to creating mounts on solaris.

stevvooe commented on snapshot/btrfs/package.go:
```diff
@@ -0,0 +1,7 @@
+package btrfs
```
Let's try to avoid importing on those platforms. We don't want the package to build if it is not useful. ;)

epilatow replied:
i added this because the root-test Makefile target expands TEST_REQUIRES_ROOT_PACKAGES using 'git grep -l testutil.RequiresRoot' and ends up including snapshot/btrfs on the "go test" command line.  given that git is not go build tag aware i sure how i could filter this entry out.  but playing around with the Makefile a bit more i figured out that if i cleanup TEST_REQUIRES_ROOT_PACKAGES i can $(filter ...) it against $(PACKAGES) to remove packages not supported on the current platform.  this allows me to delete snapshot/btrfs/package.go and snapshot/overlay/package.go.

epilatow commented:
i've updated my PR to try and address the code review comments above.
i've also re-synced with the master branch to resolve conflicts.
finally, i've added two new commits to the PR to update two vendored
components, containerd/continuity and containerd/console.  i used
github.com/LK4D4/vndr to update the vendored components.

the containerd/continuity update simply updates that component to the
latest version that now contains solaris build support.

the containerd/console update uses code from my fork of the
containerd/console repo: http://github.com/epilatow/console.  so this PR
shouldn't be integrated until my containerd/console changes are
integrated upstream and i'm able to revise this PR to use code from the
standard containerd/console repo.  i've included the code from my repo
in this PR to enable CI regression testing.

tophj-ibm commented on Makefile:
```diff
@@ -47,7 +52,11 @@ GO_TAGS=$(if $(BUILDTAGS),-tags "$(BUILDTAGS)",)
 
 # Flags passed to `go test`
-TESTFLAGS ?=-parallel 8 -race -v
+ifeq ("$(GOOS)", "solaris")
```
According to go test ``go test: -race and -msan are only supported on linux/amd64, freebsd/amd64, darwin/amd64 and windows/amd64`` So this is broken on a lot more than solaris.

Would you be willing to change this to only do -race on those combinations? 😄

epilatow replied:
sure.

epilatow commented:
i've updated my PR to try and address the code review comments above.
i've also re-synced with the master branch.

mlaventure commented on cmd/containerd/config_solaris.go:
```diff
@@ -0,0 +1,21 @@
+		},
+		Debug: server.Debug{
+			Level:   "info",
+			Address: "/var/run/containerd/debug.sock",
```
please add a `DefaultDebugAddress` to match other platforms.

epilatow replied:
sure.  thanks for pointing this out, because it also made me realize that i should be getting DefaultAddress from the server package instead of the containerd package.  so i'll fix that as well.

mlaventure commented on container_solaris_test.go:
```diff
@@ -0,0 +1,62 @@
+	"testing"
+)
+
+func TestContainerUpdate(t *testing.T) {
```
This doesn't do any update of the container resources at all. If that's not supported on Solaris, no point on having a test

epilatow replied:
will remove.

mlaventure commented on fs/copy_solaris.go:
```diff
@@ -0,0 +1,63 @@
+package fs
```
This files looks identical to `fs/copy_unix.go`

epilatow replied:
the difference is that darwin and freebsd define stat_t fields like Atimespec,
where as linux and solaris define stat_t fields like Atim.  i'll create accessor functions to hide these platform differences and then i can use copy_unix.go on solaris.

mlaventure commented on fs/dtype_test.go:
```diff
@@ -0,0 +1,18 @@
+
+func TestRequiresRootNOP(t *testing.T) {
+
+	// This is a dummy test case that exist to call
```
I don't understand, why would it fails if the function is never called?

epilatow replied:
without this running "gmake root-coverage" fails on non-linux platforms with the following "go test" error:

    flag provided but not defined: -test.root

the issue is that the makefile determines which tests to run (TEST_REQUIRES_ROOT_PACKAGES) by doing "git grep -l testutil.RequiresRoot".  in the fs package the only tests that references testutil.RequiresRoot is in dtype_linux_test.go, which is only built on linux.  but grep is not go build tag aware, so when the makefile root-coverage target runs "go test ... -test.root" on non-linux platforms, it incorrectly includes the fs package on the command line.  this fails on non-linux platforms with the error message above because there are no fs package tests that reference testutil.RequiresRoot.  i didn't see any easy way to fix the makefile to be go build tag aware, so to work around the problem i added this file to create a reference to testutil.RequiresRoot so that the "go test" commands runs successfully.

i'll try expanding the comment in this file to explain this more clearly.

if there's a better way to solve this problem then please let me know.

mlaventure replied:
Oh I see. I don't think that's still the case now, the `test.root` flag is defined regardless of the platform in `testutils/helpers.go`.

epilatow replied:
i'm currently synced up with master and using go 1.8.  if i remove fs/dtype_test.go file then i see the following failure:

    # go test github.com/containerd/containerd/fs -test.root
    flag provided but not defined: -test.root
    ....

i agree with you that testutils references test.root on all platforms.

but afaik the problem is that the only reference to the testutils package from the fs package is in dtype_linux_test.go.  so on non-linux platforms, the fs package doesn't reference the testutils package, so it doesn't define test.root.  i added fs/dtype_test.go to ensure that the fs package references the testutils package for all platforms.  that said, i'm new to go so my analysis may be incorrect or i may be overlooking a better way to fix this.

mlaventure replied:
@epilatow no, your analysis looks correct to me. Thanks!

What do you think about moving the flag registration to `client_test.go`?

Either way, the dummy test will do for now.

epilatow replied:
i presume you mean github.com/containerd/containerd/client_test.go?  i haven't tried this, but afaik none of the files in github.com/containerd/containerd/fs import github.com/containerd/containerd, so i'm not sure that would work.

epilatow commented:
i've updated my PR to try and address the code review comments above.  i've also re-synced with the master branch.  as a reminder, this PR should not be integrated until after my containerd/console PR is integrated and i vndr those changes into this PR.

crosbymichael commented:
@epilatow can you update this with the console change we just merged?

epilatow commented:
i'm currently re-syncing with master, vndr'ing x/sys to the latest, vndr'ing in the console changes, and doing regression testing.  i'll update this PR asap.

epilatow commented:
i've updated the PR with the changes mentioned above.

crosbymichael commented:
LGTM

@stevvooe wanna take a look at the fs stuff?

AkihiroSuda commented on mount/mount_solaris.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+const (
+	mountCmd = "/usr/sbin/mount"
```
Why not use `mount(2)` syscall?

epilatow replied:
the following comment at the top of the file attempts to explain this:

+// On Solaris we can't invoke the mount system call directly.  First,
 +// the mount system call takes more than 6 arguments, and go doesn't
 +// support invoking system calls that take more than 6 arguments.  Past
 +// that, the mount system call is a private interfaces.  For example,
 +// the arguments and data structures passed to the kernel to create an
 +// nfs mount are private and can change at any time.  The only public
 +// and stable interface for creating mounts on Solaris is the mount.8
 +// command, so we'll invoke that here.

AkihiroSuda replied:
ok, thanks

epilatow commented:
i resynced with the master branch to fix some merge conflicts, but now i'm seeing new CI test failures, i'm not sure why yet.

stevvooe commented on stat/stat_unix1.go:
```diff
@@ -0,0 +1,19 @@
+// +build linux solaris
+
+package stat
```
I don't think this warrants a separate package.

epilatow replied:
i made this a separate package because when i removed the code duplication in the fs package due to this issue, i noticed that there was also code duplication in the content/local due to the same issue. so what would you like me to do?  i could:
- duplicate these two files into both the fs and content/local packages.
- leave the existing duplication in the content/local package alone and just include these two files in the fs package.

stevvooe replied:
Just leave the existing duplication alone.

Sorry for being particular here, but this kind of package can have the tendency to become problematic over time.

stevvooe replied:
Another option is to put it in `sys`.

epilatow replied:
np.  moving it to sys sounds perfect to me.

stevvooe commented:
LGTM after removing the `stat` package.

epilatow commented:
i've updated the PR after moving the Stat[ACM]time interfaces into the sys packages.

mlaventure commented on sys/stat_unix2.go:
```diff
@@ -0,0 +1,19 @@
+// +build darwin freebsd
```
maybe call it `stat_bsd.go` ?

crosbymichael commented:
Pulling this down and trying to build I run into issues with `sudo make install`

```
michael|~/development/gocode/src/github.com/containerd/containerd > make && sudo make install && sudo systemctl restart containerd
🇩 bin/ctr
🇩 bin/containerd
🇩 bin/containerd-shim
🇩 binaries
make: go: Command not found
make: go: Command not found
🇩 install bin/ctr bin/containerd bin/containerd-shim
michael|~/development/gocode/src/github.com/containerd/containerd > make && sudo make install && sudo systemctl restart containerd
🇩 bin/ctr
🇩 bin/containerd
🇩 bin/containerd-shim
🇩 binaries
make: go: Command not found
make: go: Command not found
🇩 install bin/ctr bin/containerd bin/containerd-shim
```

The install command should not require go and something must have changed in this PR.

epilatow commented:
the "make install" errors seen above were caused by the following code
that i added to the Makefile:

+TESTFLAGS_RACE=
+ifeq ($(filter \
+    linux/amd64 freebsd/amd64 darwin/amd64 windows/amd64, \
+    $(GOOS)/$(GOARCH)),$(GOOS)/$(GOARCH))
+       TESTFLAGS_RACE= -race
+endif

the reference to GOARCH triggered the following variable expansion:

    GOARCH ?= $(shell go env GOARCH)

which generated the errors.  so this error would actually be seen for
any makefile target that is invoked if go is not in the PATH.  the fix
is pretty strait forward.  the Makefile already has code to check if go
is in the PATH when setting GOOS, so we can just set GOARCH at the same
time.

thanks for catching this.  i've updated the PR to fix this issue (and address
@mlaventure last nit).

crosbymichael commented:
LGTM

crosbymichael commented:
@epilatow thanks!

epilatow commented:
woot!  thanks for all the help getting this integrated.  :)
